tools/export/sw4stm32/__init__.py

```diff
@@ -362,6 +369,15 @@ def generate(self):
 
         lib_dirs = [self.filter_dot(s) for s in self.resources.lib_dirs]
 
+        preproc_cmd = ""
+        # Hack for Windows. Build fails if command contains parentheses.
+        if ('(' in self.toolchain.preproc[0] or ')' in self.toolchain.preproc[0]) and self.toolchain.preproc[0][0] != "'":
```

Comment:
Why no do this unconditionally? what would it hurt?

Reply:
Also, could you shorten this line? it's really long.

Reply:
Oh. This is wrong. There should be double quotes.

Maybe some day double quotes will be added by the toolchain class. Double double quotes would hurt.

And this will happen, if quotes are used in Linux environment:
```
"/home/jaska/eclipse/SystemWorkbench/plugins/fr.ac6.mcu.externaltools.arm-none.linux64_1.13.1.201707031232/tools/compiler/bin/arm-none-eabi-cpp -E -P" -DMBED_DEBUG -DMBED_TRAP_ERRORS_ENABLED=1 -Wl,-n -Wl,--start-group -lstdc++ -lsupc++ -lm -lc -lgcc -lnosys -Wl,--end-group ../mbed-os/targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L476xG/device/TOOLCHAIN_GCC_ARM/STM32L476XX.ld -o linker-script-debug.ld
/bin/sh: 1: /home/jaska/eclipse/SystemWorkbench/plugins/fr.ac6.mcu.externaltools.arm-none.linux64_1.13.1.201707031232/tools/compiler/bin/arm-none-eabi-cpp -E -P: not found
../makefile.targets:12: recipe for target 'linker-script-debug.ld' failed
make: *** [linker-script-debug.ld] Error 127
```

Of course I can check if environment is win32, and add quotes then.

I have tuned my line length to 120 characters. Default value is quite small for modern displays. Wait... this line is 123 characters. Automatic formatter is not working correctly here.

Reply:
>Of course I can check if environment is win32,

No you can't. Here is an example where you cannot detect the user's OS: website export. On the website, you are running Linux, but you have no way to tell what OS the user will run the export in after they have the zip file.

Therefore, you should not rely on OS detection in an exporter. 

Further, 
```
"/home/jaska/eclipse/SystemWorkbench/plugins/fr.ac6.mcu.externaltools.arm-none.linux64_1.13.1.201707031232/tools/compiler/bin/arm-none-eabi-cpp -E -P"
```

This is clearly quoted incorrectly. It should be:
```
"/home/jaska/eclipse/SystemWorkbench/plugins/fr.ac6.mcu.externaltools.arm-none.linux64_1.13.1.201707031232/tools/compiler/bin/arm-none-eabi-cpp" -E -P
```
**That** will work on Linux.

Reply:
This also points to another problem: you are using the configured toolchains. Again, in the website export situation this could lead to junk exports that contain paths that are _never_ valid for a user as they contain implementation details of the website.

Reply:
This leads to another problem. Debug configuration cannot be created, if platform is not known. Is it possible to detect, if export was called by the website?

Reply:
I forgot that preproc_cmd is not a plain command, but also has two command line argument.

Toolchain is just a last resort to get preprocessor command. And value comes from the GCC class:
`self.preproc = [join(tool_path, "arm-none-eabi-cpp"), "-E", "-P"]`
If path is not set, value is plain arm-none-eabi-cpp command with arguments. That should also work with SW4STM32.

Reply:
>  Is it possible to detect, if export was called by the website?

No.

> if path is not set, value is plain arm-none-eabi-cpp command with arguments. That should also work with SW4STM32.

Yup. 

Knowing this, I recommend wiping out the path to the executable, and just having an invocation like `arm-none-eabi-cpp` like the makefile exporters do.